Fix MG Transfer for Tetrahedrons

dominiktassilostill opened the pull request:
Follow up to #16856. As discussed briefly in #16905, this PR uses the infrastructure introduced to choose the prolongation and restriction operators according to the refinement choice used to refine a tetrahedron. Therefore, the transpose of the projection matrix is used for restriction.
I'm not sure yet how to test this properly, any recommendations are welcome! Further, I'm not too familiar how to handle the MPI communication in `BlackBoxFineDoFHandlerView::reinit()` or if it is even necessary in `BlackBoxFineDoFHandlerView::get_cell_view()` to know also about the refinement choices of the remote children.

peterrum commented:
> Further, I'm not too familiar how to handle the MPI communication in BlackBoxFineDoFHandlerView::reinit() or if it is even necessary in BlackBoxFineDoFHandlerView::get_cell_view() to know also about the refinement choices of the remote children.

Looks fine. You are treating `refinement_case` as `active_fe_index` :+1: Is it not working?

How are we storing the `refinement_case` per child or parent? All children of a parent have the same one so that it would make sense to have it there. Independent, we should probably add a check that the `refinement_case`s of all children are the same.

dominiktassilostill commented:
The `refinement_case` is stored with the parent, as it is the cell which is refined with a specific choice. The parent is also used to get the information from, I changed it such that it now gets it from `cell_coarse_on_fine_mesh` directly and not from the children on the fine mesh. Also added a check to see if the returned `refinement_case` is the same for all children. 
`RefinementCase<dim>::no_refinement` is the standard value on the cell, it is used to double check, if the cell really was refined (the other check is happening in `process_cells`).

peterrum commented on tests/multigrid-global-coarsening/mg_transfer_a_06.cc:
```diff
@@ -0,0 +1,326 @@
+/**
+ * Test prolongation and restriction with uniformly refined meshes.
+ *
+ * To compare with fe/up_and_down_simplex.
```
What does the test actually do?

dominiktassilostill replied:
I used it as a sanity check to see if the operations done are reasonable, but it does not add any additional value compared to `mg_transfer_a_05` apart from testing the `interpolate()` function. So maybe just remove this?

kronbichler replied:
So the test for `interpolate()` is not done elsewhere? In that case, I would suggest to keep that part as a separate test, name the test case accordingly (or write the comment here on what the special purpose of this test is). I like to have tests that are specific, so in case something fails I can look into the test and see immediately what the special purpose of this particular test is.

dominiktassilostill replied:
Changed the test to only test the `interpolate()` function and added a comment about it.

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -893,6 +934,10 @@ namespace internal
+                if (cell->level() != 0)
+                  buffer.push_back(cell->parent()->refinement_case());
+                else
+                  buffer.push_back(numbers::invalid_unsigned_int);
```
Is `invalid_unsigned_int` the right number here? I'm asking because `buffer` storing the argument passed in is of type `types::global_dof_index` and `refinement_case` is of type `std::uint8_t`. I would pick either `types::invalid_dof_index` or `static_cast<std::unit8_t>(-1)`.

dominiktassilostill replied:
Changed to `numbers::invalid_dof_index`.

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -962,13 +1007,16 @@ namespace internal
             for (unsigned int i = 0, k = 0; i < ids.size(); ++i)
               {
                 const unsigned int active_fe_index = buffer[k++];
+                const unsigned int refinement_case = buffer[k++];
```
Similar to my comment above, do you want to make this reflect the type intended for `refinement_case`, i.e., `std::unit8_t`?

dominiktassilostill replied:
The compiler complained when casting straight to `std::unit8_t`, but calling `RefinementCase()` first also seems to fix it.

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -1124,8 +1204,10 @@ namespace internal
-      map;
+             std::tuple<unsigned int,
+                        std::vector<types::global_dof_index>,
+                        unsigned int>>
```
same as above,
```suggestion
                        std::uint8_t>>
```

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -1680,43 +1766,74 @@ namespace internal
       // continuous or discontinuous
-      transfer.fine_element_is_continuous = fe_fine.n_dofs_per_vertex() > 0;
+      transfer.fine_element_is_continuous  = fe_fine.n_dofs_per_vertex() > 0;
+      unsigned int current_refinement_case = numbers::invalid_unsigned_int;
```
```suggestion
      std::uint8_t current_refinement_case = static_cast<std::uint8_t>(-1);
```
and similarly in a few places below.

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -1778,18 +1895,20 @@ namespace internal
-        unsigned int cell_no_0 = 0;
-        unsigned int cell_no_1 = transfer.schemes[0].n_coarse_cells;
+        unsigned int n_coarse_cells_total = 0;
+        for (unsigned int i = 0; i < transfer.schemes.size(); i++)
+          {
+            n_coarse_cells_total += transfer.schemes[i].n_coarse_cells;
+          }
```
Maybe a range-based for loop is more elegant here (otherwise, we would also prefer using prefix increment operators `++i` in deal.II over the postfix variant):
```suggestion
        for (const auto &scheme : transfer.schemes)
          n_coarse_cells_total += scheme.n_coarse_cells;
```

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -1824,51 +1947,56 @@ namespace internal
             // move pointers
             {
-              cell_no_0++;
+              cell_no[0]++;
```
This is preexisting, but would we also want to use prefix increments,
```suggestion
              ++cell_no[0];
```

kronbichler reviewed (approved):
Apart from the comments by @peterrum, I have a few comments regarding consistency of data types. I think this looks really good, it is a great step forward :+1:

kronbichler commented:
There is a failure on the `MPI` test machine for the test `multigrid-global-coarsening/transfer_bug_01.mpirun=2.debug`:
```
An error occurred in line <2863> of file </jenkins/workspace/dealii-mpi_PR-16959/include/deal.II/base/geometry_info.h> in function
    dealii::RefinementCase<dim>::RefinementCase(uint8_t) [with int dim = 2; uint8_t = unsigned char]
The violated condition was: 
    (refinement_case & RefinementPossibilities<dim>::isotropic_refinement) == refinement_case
Additional information: 
    The refinement flags given (255) contain set bits that do not make
    sense for the space dimension of the object to which they are applied.
```

dominiktassilostill commented:
> There is a failure on the `MPI` test machine for the test `multigrid-global-coarsening/transfer_bug_01.mpirun=2.debug`

There is no refinement in the test, so all cells are on level `0`, leading to only `numbers::invalid_dof_index` being communicated. This number then is an invalid argument for `RefinementCase()` (which is used to convert it into an `unsinged char` to keep the data types consistent), instead `static_cast<std::uint8_t>()` is now called to convert it. 
I assume this kind of grid setup will always lead to a `IdentityTransfer`? Else some asserts later in the code will be triggered.

kronbichler commented on doc/news/changes/major/20240510Still:
```diff
@@ -0,0 +1,3 @@
+Updated: Tetrahedral meshes are now refined quasi-uniformly, preventing degradation with higher refinement levels.
```
I would add that also the multigrid infrastructure has been updated to take this into account:
```suggestion
Updated: Tetrahedral meshes are now refined quasi-uniformly, preventing degradation
with higher refinement levels. The changes also include the relevant adjustments in the
geometric multigrid infrastructure.
```
I think it might make sense to unify this with the previous entry https://github.com/dealii/dealii/blob/6495265e5800149683b2ce3f002ba60f64031bff/doc/news/changes/minor/20240320StillFehnKronbichler#L1-L8
I would personally remove that old entry in favor of the new one, not as a way to diminish the contribution but just for clarity to the users not involved in these changes. However, I think we should add another "minor" entry regarding the correction for the restriction matrices you did for `FE_SimplexP`.
